cocotb/share/lib/vpi/VpiImpl.cpp

```diff
@@ -296,13 +296,46 @@ GpiObjHdl* VpiImpl::native_check_create(void *raw_hdl, GpiObjHdl *parent)
 GpiObjHdl* VpiImpl::native_check_create(std::string &name, GpiObjHdl *parent)
 {
     vpiHandle new_hdl;
+    const vpiHandle parent_hdl = parent->get_handle<vpiHandle>();
     std::string fq_name = parent->get_fullname() + "." + name;
-    std::vector<char> writable(fq_name.begin(), fq_name.end());
-    writable.push_back('\0');
 
-    new_hdl = vpi_handle_by_name(&writable[0], NULL);
+    new_hdl = vpi_handle_by_name(const_cast<char*>(fq_name.c_str()), NULL);
+
+#ifdef ICARUS
+    /* Icarus does not support vpiGenScopeArray, only vpiGenScope.
+     * If not found by name, look for generate block with matching
+     * name.
+     *     For Example:
+     *         genvar idx;
+     *         generate
+     *             for (idx = 0; idx < 5; idx = idx + 1) begin
+     *                 ...
+     *             end
+     *         endgenerate
+     *
+     *     Icarus finds genblk1[0], so genblk1 can be inferred.
+     */
+    if (new_hdl == NULL) {
+        vpiHandle iter = vpi_iterate(vpiInternalScope, parent_hdl);
+        if (iter == NULL) {
+            goto skip_iterate;
+        }
+
+        vpiHandle rgn;
+        for (rgn = vpi_scan(iter); rgn != NULL; rgn = vpi_scan(iter)) {
+            if (vpi_get(vpiType, rgn) == vpiGenScope) {
+                std::string rgn_name = vpi_get_str(vpiName, rgn);
+                if (rgn_name.compare(0, name.length(), name) == 0) {
```

Comment:
Just use `rgn_name == name`. I'd suggest you leave `rgn_name` as a `const char*` using `auto` to avoid an allocation.

```suggestion
                auto rgn_name = vpi_get_str(vpiName, rgn);
                if (rgn_name == name) {
```

Reply:
Will that work?
Example:
`name` is "genblk"
`rgn_name` is "genblk[0]"

We want to check that `rgn_name` *starts with* `name`. `compare()` without *pos* and *len* parameters would return >0 I think.

Reply:
Ah, that wasn't immediately apparent to me that's what this code was doing (maybe warrants a comment or a `startswith` function?). You can still avoid the allocation though https://en.cppreference.com/w/cpp/string/basic_string/compare overload 5.

Reply:
Overload 5 `name.compare(0, std::string::npos, rgn_name)` fails as it compares to the end of the longer null-terminated `rgn_name`.
Resorted to `std::strncmp` unless you have another idea.

Reply:
My bad overload 6 would work.